src/toolkits/drawing_classifier/data_preparation.cpp

```diff
@@ -230,84 +231,132 @@ flex_list simplify_drawing(flex_list raw_drawing, int row_number) {
     return simplified_drawing;
 }
 
-bool in_bounds(int x, int y, int dim) {
-    return (x >= 0 && x < dim && y >= 0 && y < dim);
+bool in_bounds(int x, int y, int x_dim, int y_dim) {
+    return (x >= 0 && x < x_dim && y >= 0 && y < y_dim);
 }
 
 void paint_point(flex_nd_vec &bitmap, int x, int y, int pad) {
-    size_t dimension = bitmap.shape()[1];
+    size_t x_dimension = bitmap.shape()[0];
+    size_t y_dimension = bitmap.shape()[1];
     for (int dx = -pad; dx < pad; dx++) {
         for (int dy = -pad; dy < pad; dy++) {
-            if (in_bounds(x+dx, y+dy, dimension)) {
-                bitmap[(y+dy) * dimension + (x+dx)] = 1.0;
+            if (in_bounds(x + dx, y + dy, x_dimension, y_dimension)) {
+                DASSERT_GE(x + dx, 0);
+                DASSERT_LT(x + dx, x_dimension);
+                DASSERT_GE(y + dy, 0);
+                DASSERT_LT(y + dy, y_dimension);
+#ifndef NDEBUG
+                std::vector<size_t> indices = {0,(size_t)(y+dy),(size_t)(x+dx)};
+                DASSERT_EQ(
+                    bitmap[(y+dy) * x_dimension + (x+dx)],
+                    bitmap.at(bitmap.fast_index(indices))
+                    );
+#endif
+                bitmap[(y + dy) * x_dimension + (x + dx)] = 1.0;
             }
         }
     }
 }
 
-flex_nd_vec paint_stroke(
-    flex_nd_vec bitmap, Point start, Point end, float stroke_width) {
-    bool along_x;
-    float slope;
-    if (floorf(end.get_x()) == floorf(start.get_x())) {
-        slope = std::numeric_limits<float>::max();
-    } else {
-        slope = (end.get_y() - start.get_y())/(end.get_x() - start.get_x());
-    }
-    int pad = (int)(stroke_width/2);
-    along_x = (fabs(slope) < 1);
-    if ((along_x && (start.get_x() > end.get_x()))
-        || (!along_x && (start.get_y() > end.get_y()))) {
-        std::swap(start, end);
+flex_nd_vec paint_stroke(flex_nd_vec &bitmap, Point &start, Point &end,
+                         float stroke_width) {
+  bool along_x;
+  float slope;
+  if (floorf(end.get_x()) == floorf(start.get_x())) {
+    slope = std::numeric_limits<float>::max();
+  } else {
+    // Make sure denominator of slope is non-zero, not necessarily positive.
+    DASSERT_NE(end.get_x() - start.get_x(), 0);
```

Comment:
Since either `end.get_x()` or `start.get_x()` could be greater than the other but the difference must be non-zero (for the division in the next line to not be undefined), we check for non-zeroness only.

Reply:
DASSERT_NE(end.get_x(), start.get_x()) would be clearer.